distributed/tests/test_core.py

```diff
@@ -994,3 +996,69 @@ async def long_handler(comm, delay=10):
         await asyncio.wait_for(fut, 0.5)
     await comm.close()
     await server.close()
+
+
+def test_expects_comm():
+    class A:
+        def empty(self):
+            ...
+
+        def one_arg(self, arg):
+            ...
+
+        def comm_arg(self, comm):
+            ...
+
+        def stream_arg(self, stream):
+            ...
+
+        def two_arg(self, arg, other):
+            ...
+
+        def comm_arg_other(self, comm, other):
+            ...
+
+        def stream_arg_other(self, stream, other):
+            ...
+
+        def arg_kwarg(self, arg, other=None):
+            ...
+
+        def comm_posarg_only(self, comm, /, other):
+            ...
+
+        def comm_kwarg_only(self, *, comm, other):
+            ...
+
+        def comm_not_leading_position(self, *, other, comm):
+            ...
+
+        def stream_not_leading_position(self, *, other, stream):
+            ...
+
+        stream_not_leading_partialmethod = partialmethod(
+            stream_not_leading_position, other="foo"
+        )
```

Comment:
```suggestion
```

Reply:
I actually think the "not leading position" thing is important since I don't want to block any names from signatures. If a user wants to define a stream kwarg or a comm kwarg without using our comm mechanism that's fine imo.

I generally dislike this API quite a lot. I don't think we should silently expect the first argument to be positional-only comm.